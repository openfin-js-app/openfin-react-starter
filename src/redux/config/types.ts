import { SvgIconProps } from '@material-ui/core/SvgIcon';
import { TextFieldProps } from '@material-ui/core/TextField';
import {TimePickerModalProps} from "material-ui-pickers/TimePicker/TimePickerModal";
import {DatePickerModalProps} from "material-ui-pickers/DatePicker/DatePickerModal";


export enum FieldType{
    CUSTOM_FIELD = 1,
    STRING = 2,
    DATETIME = 3,
    DATE = 4,
    TIME = 5,
    NUMBER = 6,
    CURRENCY = 7,
    TITLE = 8,
    SUBHEADING = 9,
    BODY1 = 10,
}

export interface IConfigField {
    _type:FieldType;
    _label:string;
    _name?:string;
    _defaultValue?:any;
    _cols?:number;
    _rows?:number;
    _props?:TextFieldProps|TimePickerModalProps|DatePickerModalProps;
    _custom?:any;
    [key:string]:any;
    [key:number]:any;
}

export interface IConfigTab {
    _order?:number;
    _label:string;
    _name:string;
    _svgUrl:string;
    _icon?:React.ComponentType<SvgIconProps>;
    _fields:IConfigField[];
    _fieldLabels?:string;
    [key:string]:any;
    [key:number]:any;
}

export interface IConfigResetOption {
    tabName?:string
}

export interface IConfigUpdateOneFieldOption {
    name:string,
    value:any,
}

export interface IConfigUpdateGlobalFilterStrOption {
    configGlobalFilterString:string,
}

export interface IConfigState {
    configGlobalFilterString:string;
    _tabs:IConfigTab[];
    [key:string]:any;
    [key:number]:any;
}