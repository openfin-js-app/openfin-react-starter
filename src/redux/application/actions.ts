import { createAction, ActionFunctionAny, Action } from 'redux-actions';
import { IApplicationNewSnackbarOption, IApplicationSetSnackbarStatusOption, IApplicationCloseSnackBarOption } from './types';

export const APPLICATION_STARTED:string = 'APPLICATION_STARTED';
export const APPLICATION_READY:string = 'APPLICATION_READY';
export const APPLICATION_DRAWER_TOGGLE:string = 'APPLICATION_DRAWER_TOGGLE';
export const APPLICATION_TOGGLE_WINDOW_STATE:string = 'APPLICATION_TOGGLE_WINDOW_STATE';

export const applicationStarted:ActionFunctionAny<Action<void>> = createAction(APPLICATION_STARTED);
export const applicationReady:ActionFunctionAny<Action<void>> = createAction(APPLICATION_READY);
export const applicationDrawerToggle:ActionFunctionAny<Action<void>> = createAction(APPLICATION_DRAWER_TOGGLE);
export const applicationToogleWindowState:ActionFunctionAny<Action<void>> = createAction(APPLICATION_TOGGLE_WINDOW_STATE);

// snackbar

export const APPLICATION_NEW_SNACKBAR='APPLICATION_NEW_SNACKBAR';
export const APPLICATION_SET_SNACKBAR_STATUS='APPLICATION_SET_SNACKBAR_STATUS';
export const APPLICATION_PROCESS_SNACKBAR_QUEUE='APPLICATION_PROCESS_SNACKBAR_QUEUE';
export const APPLICATION_CLOSE_SNACKBAR='APPLICATION_CLOSE_SNACKBAR';

export const applicationNewSnackbar:ActionFunctionAny<Action<IApplicationNewSnackbarOption>>
    = createAction(APPLICATION_NEW_SNACKBAR, (option:IApplicationNewSnackbarOption)=>(option));
export const applicationSetSnackbarStatus:ActionFunctionAny<Action<IApplicationSetSnackbarStatusOption>>
    = createAction(APPLICATION_SET_SNACKBAR_STATUS, (option:IApplicationSetSnackbarStatusOption)=>(option));
export const applicationProcessSnackbarQueue:ActionFunctionAny<Action<void>>
    = createAction(APPLICATION_PROCESS_SNACKBAR_QUEUE);
export const applicationCloseSnackbar:ActionFunctionAny<Action<IApplicationCloseSnackBarOption>>
    = createAction(APPLICATION_CLOSE_SNACKBAR, (option)=>(option));


// launch bar

export const APPLICATION_LAUNCH_BAR_TOGGLE = 'APPLICATION_LAUNCH_BAR_TOGGLE';
export const APPLICATION_LAUNCH_BAR_TOGGLE_COLLAPSE = 'APPLICATION_LAUNCH_BAR_TOGGLE_COLLAPSE';
export const APPLICATION_LAUNCH_NEW_WINDOW = 'APPLICATION_LAUNCH_NEW_WINDOW';
export const applicationLaunchBarToggle:ActionFunctionAny<Action<void>> = createAction(APPLICATION_LAUNCH_BAR_TOGGLE);
export const applicationLaunchBarToggleCollapse:ActionFunctionAny<Action<void>> = createAction(APPLICATION_LAUNCH_BAR_TOGGLE_COLLAPSE);
export const applicationLaunchNewWindow:ActionFunctionAny<Action<any>>
    = createAction(APPLICATION_LAUNCH_NEW_WINDOW,(appJson:any)=>(appJson));