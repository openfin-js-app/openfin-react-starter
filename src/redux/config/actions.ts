import {createAction, ActionFunctionAny, Action} from 'redux-actions';

import {IConfigResetOption, IConfigUpdateOneFieldOption, IConfigUpdateGlobalFilterStrOption} from './types';

export const CONFIG_RESET:string = 'CONFIG_RESET';
export const CONFIG_UPDATE_ONE_FIELD:string = 'CONFIG_UPDATE_ONE_FIELD';
export const CONFIG_UPDATE_GLOBAL_FILTER_STR:string = 'CONFIG_UPDATE_GLOBAL_FILTER_STR';

export const configReset:ActionFunctionAny<Action<IConfigResetOption>>
    = createAction(CONFIG_RESET, (option)=>(option));
export const configUpdateOneField:ActionFunctionAny<Action<IConfigUpdateOneFieldOption>>
    = createAction(CONFIG_UPDATE_ONE_FIELD, (option)=>(option));
export const configUpdateGlobalFilterStr:ActionFunctionAny<Action<IConfigUpdateGlobalFilterStrOption>>
    = createAction(CONFIG_UPDATE_GLOBAL_FILTER_STR, (option)=>(option));

// optional actions depending on the config application values

export const CONFIG_UPDATE_NEW_WINDOW_POSITION:string = 'CONFIG_UPDATE_NEW_WINDOW_POSITION';
export const CONFIG_UPDATE_NEW_WINDOW_POSITION_ADD_DELTA:string = 'CONFIG_UPDATE_NEW_WINDOW_POSITION_ADD_DELTA';
export const CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_TOP:string = 'CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_TOP';
export const CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_LEFT:string = 'CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_LEFT';
export const configUpdateNewWindowPosition:ActionFunctionAny<Action<void>>
    = createAction(CONFIG_UPDATE_NEW_WINDOW_POSITION);
export const configUpdateNewWindowPositionAddDelta:ActionFunctionAny<Action<void>>
    = createAction(CONFIG_UPDATE_NEW_WINDOW_POSITION_ADD_DELTA);
export const configUpdateNewWindowPositionResetTop:ActionFunctionAny<Action<void>>
    = createAction(CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_TOP);
export const configUpdateNewWindowPositionResetLeft:ActionFunctionAny<Action<void>>
    = createAction(CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_LEFT);