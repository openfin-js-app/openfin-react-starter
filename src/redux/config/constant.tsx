import * as React from 'react';
import InputAdornment from '@material-ui/core/InputAdornment';

import { ConfigTab, ConfigField, FieldType } from './types';

import SettingIcon from '@material-ui/icons/Settings';

const controls = require('../../assets/svg/support/controls_dark.svg') as string;

const configTabs:ConfigTab[]=[
    {
        _order: 0 ,
        _label: 'Application',
        _name:'application',
        _svgUrl: controls,
        _icon: SettingIcon,
        _fields:[
            {
                _type:FieldType.SUBHEADING,
                _label:'New Window Position',
                _cols: 12,
            },
            {
                _type:FieldType.NUMBER,
                _label:'New Window Top',
                _name:'newWinTop',
                _defaultValue:parseInt(process.env.REACT_APP_NEW_WINDOW_TOP),
                _props:{
                    InputProps:{
                        endAdornment:<InputAdornment position={'end'}>Pixel</InputAdornment>
                    }
                }
            },
            {
                _type:FieldType.NUMBER,
                _label:'New Window Left',
                _name:'newWinLeft',
                _defaultValue:parseInt(process.env.REACT_APP_NEW_WINDOW_LEFT),
                _props:{
                    InputProps:{
                        endAdornment:<InputAdornment position={'end'}>Pixel</InputAdornment>
                    }
                }
            },
            {
                _type:FieldType.NUMBER,
                _label:'New Window Width',
                _name:'newWinWidth',
                _defaultValue:parseInt(process.env.REACT_APP_NEW_WINDOW_WIDTH),
                _props:{
                    InputProps:{
                        endAdornment:<InputAdornment position={'end'}>Pixel</InputAdornment>
                    }
                }
            },
            {
                _type:FieldType.NUMBER,
                _label:'New Window Height',
                _name:'newWinHeight',
                _defaultValue:parseInt(process.env.REACT_APP_NEW_WINDOW_HEIGHT),
                _props:{
                    InputProps:{
                        endAdornment:<InputAdornment position={'end'}>Pixel</InputAdornment>
                    }
                }
            },
            {
                _type:FieldType.NUMBER,
                _label:'New Window Delta Left',
                _name:'newWindDeltaLeft',
                _defaultValue:parseInt(process.env.REACT_APP_NEW_WINDOW_DELTA_LEFT),
                _props:{
                    InputProps:{
                        endAdornment:<InputAdornment position={'end'}>Pixel</InputAdornment>
                    }
                }
            },
            {
                _type:FieldType.NUMBER,
                _label:'New Window Delta Height',
                _name:'newWindDeltaHeight',
                _defaultValue:parseInt(process.env.REACT_APP_NEW_WINDOW_DELTA_HEIGHT),
                _props:{
                    InputProps:{
                        endAdornment:<InputAdornment position={'end'}>Pixel</InputAdornment>
                    }
                }
            },
        ]
    }
];

export default configTabs;
