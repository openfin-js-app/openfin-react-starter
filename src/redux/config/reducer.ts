import { handleActions, Action } from 'redux-actions';

import {
    CONFIG_RESET, CONFIG_UPDATE_ONE_FIELD,
    CONFIG_UPDATE_GLOBAL_FILTER_STR,

    CONFIG_UPDATE_NEW_WINDOW_POSITION_ADD_DELTA,
    CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_TOP,
    CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_LEFT,
} from './actions';
import { ConfigTab, ConfigField, FieldType, ConfigState } from './types';
import configTabs from './constant';

function buildDefaultState(configTabs: ConfigTab[]):ConfigState{
    let result:ConfigState = {
        configGlobalFilterString:'',
        _tabs:configTabs,
    };
    configTabs.forEach(oneTab =>{
        oneTab._fieldLabels='';
        result[oneTab._name]={};
        oneTab._fields.forEach(oneField =>{
            oneTab._fieldLabels = oneTab._fieldLabels+oneField._label;
            if(oneField._name && (oneField._defaultValue || oneField._defaultValue==="")){
                result[oneTab._name][oneField._name]=oneField._defaultValue;
            }
        })
    });
    return result;
}

const defaultState:ConfigState = buildDefaultState(configTabs);

export default handleActions({
    [CONFIG_RESET]:(state,action)=>{
      const {tabName} = action.payload;
      if(tabName){
          return {
              ...state,
              tabName:defaultState[tabName],
          }
      }else{
          return defaultState;
      }
    },
    [CONFIG_UPDATE_ONE_FIELD]:(state,action)=>{
        const {name,value} = action.payload;
        let result = {...state};
        let paths = name.split('.');
        if (paths.length == 2){
            result[paths[0]]={
                ...result[paths[0]],
                [paths[1]]:value,
            }
        }
        return result;
    },
    [CONFIG_UPDATE_GLOBAL_FILTER_STR]:(state,action)=>{
        const { configGlobalFilterString } = action.payload;
        return{
            ...state,
            configGlobalFilterString,
        };
    },
    [CONFIG_UPDATE_NEW_WINDOW_POSITION_ADD_DELTA]:(state,action)=>{
        const applicationConfig = state.application;

        return {
            ...state,
            application:{
                ...state.application,
                newWinTop: applicationConfig.newWinTop+ applicationConfig.newWindDeltaHeight,
                newWinLeft: applicationConfig.newWinLeft+ applicationConfig.newWindDeltaLeft,
            }
        };
    },
    [CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_TOP]:(state,action)=>({
        ...state,
        application:{
            ...state.application,
            newWinTop: parseInt(process.env.REACT_APP_NEW_WINDOW_TOP),
        }
    }),
    [CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_LEFT]:(state,action)=>({
        ...state,
        application:{
            ...state.application,
            newWinLeft: parseInt(process.env.REACT_APP_NEW_WINDOW_LEFT),
        }
    }),
},defaultState);