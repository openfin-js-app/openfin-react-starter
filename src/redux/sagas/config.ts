import {put, select, take, takeLatest,} from 'redux-saga/effects';

import {
    CONFIG_UPDATE_NEW_WINDOW_POSITION,
    configUpdateNewWindowPositionAddDelta,
    configUpdateNewWindowPositionResetLeft,
    configUpdateNewWindowPositionResetTop,
} from '..';
import {System, Window} from "@albertli90/redux-openfin";

export const getNewWindowTop = state => state.config.application.newWinTop;
export const getNewWindowLeft = state => state.config.application.newWinLeft;
export const getNewWindowWidth = state => state.config.application.newWinWidth;
export const getNewWindowHeight = state => state.config.application.newWinHeight;

export function* handleConfigUpdateNewWindowPosition() {
    const newWinWidth = yield select(getNewWindowWidth);
    const newWinHeight = yield select(getNewWindowHeight);
    const newWinTop = yield select(getNewWindowTop);
    const newWinLeft = yield select(getNewWindowLeft);

    yield put.resolve(System.actions.getMonitorInfo({}));
    const monitorInfoAction = yield take(System.actions.GET_MONITOR_INFO_RES);
    const virtualScreen = monitorInfoAction.payload.virtualScreen;

    // console.log("configUpdateNewWindowPosition",monitorInfoAction,virtualScreen);

    if (
        ((newWinLeft+newWinWidth)<virtualScreen.right) &&
        ((newWinTop+newWinHeight)<virtualScreen.bottom)
    ){
        yield put.resolve(configUpdateNewWindowPositionAddDelta());
    }else{
        if ((newWinLeft+newWinWidth)>=virtualScreen.right){
            yield put.resolve(configUpdateNewWindowPositionResetLeft());
        }
        if ((newWinTop+newWinHeight)>=virtualScreen.bottom){
            yield put.resolve(configUpdateNewWindowPositionResetTop());
        }
    }

}

export default function* () {
    yield takeLatest(CONFIG_UPDATE_NEW_WINDOW_POSITION,handleConfigUpdateNewWindowPosition);
}