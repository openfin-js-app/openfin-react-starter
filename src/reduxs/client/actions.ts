import { createAction, ActionFunctionAny, Action } from 'redux-actions';
import { IClientState } from './types';

export const CLIENT_SET_VALUE:string = 'CLIENT_SET_VALUE';
export const clientSetValue:ActionFunctionAny<Action<Partial<IClientState>>> = createAction(CLIENT_SET_VALUE,
    (option:Partial<IClientState>)=>(option)
);