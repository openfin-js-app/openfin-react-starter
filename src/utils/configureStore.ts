import { applyMiddleware, createStore, compose } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { createOpenfinMiddleware } from '@albertli90/redux-openfin';
import { ChannelType } from '@albertli90/redux-openfin/init';

import rootReducer, {IRootState} from '../reduxs';
import rootSaga from '../reduxs/sagas';

declare const window:any;

export default (
        channelType:ChannelType,
        channelClientId:string,
        sharedActions:string[],
        parentState?:IRootState
)=>{

    const openfinMiddleware = createOpenfinMiddleware(window.fin,{
        channelType,channelClientId,sharedActions,
        channelRandomSuffix:process.env.NODE_ENV === 'development',
        autoDocking:process.env.REACT_APP_ENABLE_AUTO_DOCKING === 'true',
        dockingOptions:{
            range: 120,
        }
    });
    const sagaMiddleware = createSagaMiddleware();
    const devtools = window.devToolsExtension?window.devToolsExtension():(f:any):any => (f);

    const middleware = compose(
        applyMiddleware(
            sagaMiddleware,
            openfinMiddleware,
        ),
        devtools
    );

    const store = createStore(
        rootReducer(parentState),
        middleware,
    );

    sagaMiddleware.run(rootSaga);

    return store;

}