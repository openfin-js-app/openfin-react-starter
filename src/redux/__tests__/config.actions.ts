import configureStore from 'redux-mock-store';
import * as Actions from '../config/actions';

const mockStore = configureStore();
const initialState = {};

const store = mockStore(initialState);

describe('Config actions',()=>{

    beforeEach(()=>{
        store.clearActions();
    });

    it('configReset created and dispatched correctly',()=>{
        store.dispatch(Actions.configReset());
        expect(store.getActions()).toMatchSnapshot();
    });

    it('configUpdateOneField created and dispatched correctly',()=>{
        store.dispatch(Actions.configUpdateOneField({
            name:'name', value:'value',
        }));
        expect(store.getActions()).toMatchSnapshot();
    });

    it('configUpdateGlobalFilterStr created and dispatched correctly',()=>{
        store.dispatch(Actions.configUpdateGlobalFilterStr({
            configGlobalFilterString:'configGlobalFilterString',
        }));
        expect(store.getActions()).toMatchSnapshot();
    });

    it('configUpdateNewWindowPosition created and dispatched correctly',()=>{
        store.dispatch(Actions.configUpdateNewWindowPosition());
        expect(store.getActions()).toMatchSnapshot();
    });

    it('configUpdateNewWindowPositionAddDelta created and dispatched correctly',()=>{
        store.dispatch(Actions.configUpdateNewWindowPositionAddDelta());
        expect(store.getActions()).toMatchSnapshot();
    });

    it('configUpdateNewWindowPositionResetTop created and dispatched correctly',()=>{
        store.dispatch(Actions.configUpdateNewWindowPositionResetTop());
        expect(store.getActions()).toMatchSnapshot();
    });

    it('configUpdateNewWindowPositionResetLeft created and dispatched correctly',()=>{
        store.dispatch(Actions.configUpdateNewWindowPositionResetLeft());
        expect(store.getActions()).toMatchSnapshot();
    });

});