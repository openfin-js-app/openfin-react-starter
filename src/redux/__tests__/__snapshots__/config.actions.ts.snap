// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config actions configReset created and dispatched correctly 1`] = `
Array [
  Object {
    "type": "CONFIG_RESET",
  },
]
`;

exports[`Config actions configUpdateGlobalFilterStr created and dispatched correctly 1`] = `
Array [
  Object {
    "payload": Object {
      "configGlobalFilterString": "configGlobalFilterString",
    },
    "type": "CONFIG_UPDATE_GLOBAL_FILTER_STR",
  },
]
`;

exports[`Config actions configUpdateNewWindowPosition created and dispatched correctly 1`] = `
Array [
  Object {
    "type": "CONFIG_UPDATE_NEW_WINDOW_POSITION",
  },
]
`;

exports[`Config actions configUpdateNewWindowPositionAddDelta created and dispatched correctly 1`] = `
Array [
  Object {
    "type": "CONFIG_UPDATE_NEW_WINDOW_POSITION_ADD_DELTA",
  },
]
`;

exports[`Config actions configUpdateNewWindowPositionResetLeft created and dispatched correctly 1`] = `
Array [
  Object {
    "type": "CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_LEFT",
  },
]
`;

exports[`Config actions configUpdateNewWindowPositionResetTop created and dispatched correctly 1`] = `
Array [
  Object {
    "type": "CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_TOP",
  },
]
`;

exports[`Config actions configUpdateOneField created and dispatched correctly 1`] = `
Array [
  Object {
    "payload": Object {
      "name": "name",
      "value": "value",
    },
    "type": "CONFIG_UPDATE_ONE_FIELD",
  },
]
`;
